Panic: index out of range in sliceColumnsByIndices during high concurrency writes

# Panic: index out of range in sliceColumnsByIndices during high concurrency writes

## Environment
- Arc version: 26.01.2
- Kubernetes, single arc-write pod

## Reproduce
1. Send high volume writes to `/api/v1/write/line-protocol`
2. 48 concurrent workers, 5000-row batches
3. Crash occurs intermittently during flush

## Error
```
panic: runtime error: index out of range [20000] with length 20000

goroutine 82 [running]:
github.com/basekick-labs/arc/internal/ingest.sliceColumnsByIndices(...)
	/build/internal/ingest/arrow_writer.go:2259
github.com/basekick-labs/arc/internal/ingest.(*ArrowBuffer).flushPartitionedData(...)
	/build/internal/ingest/arrow_writer.go:1775
```

## Root Cause

Race condition in `flushBufferLocked()` (`arrow_writer.go:1817-1870`):

```go
// Line 1851: Lock released before merge
shard.mu.Unlock()

// Line 1854: Merge batches creates column arrays
merged, err := b.mergeBatches(recordsToFlush)

// Line 1862: Flush uses these arrays
b.flushBufferLockedDataTime(...)
```

In `flushPartitionedData()`:
- `groupByHour(times)` creates indices `0..N-1` based on `times` length
- `sliceColumnsByIndices(merged, bucket.indices)` accesses other columns using these indices
- If another column has different length due to race → panic

## Suggested Fix

Add bounds check in `sliceColumnsByIndices()`:
```go
for i, idx := range indices {
    if idx >= len(col) {
        return nil, fmt.Errorf("index %d out of bounds for column %s (len=%d)", idx, colName, len(col))
    }
    newCol[i] = col[idx]
}
```

Or validate all columns have equal length after `mergeBatches()`.

--- a/RELEASE_NOTES_2026.02.1.md
+++ b/RELEASE_NOTES_2026.02.1.md
@@ -231,6 +231,45 @@ Fixed SSL certificate validation errors when querying data from Azure Blob Stora
 
 *Contributed by [@schotime](https://github.com/schotime)*
 
+### UTC Consistency for Compaction Filenames (PR #132)
+
+Fixed timezone inconsistency in compacted file naming. Files were being named using the server's local timezone instead of UTC, which could cause issues when servers in different timezones processed the same data.
+
+**Changes:**
+- Compacted filenames now use UTC timestamps consistently (`time.Now().UTC()`)
+- Updated daily file parsing to handle the full timestamp format (`YYYYMMDD_HHMMSS`)
+- Removed unused `GetCompactedFilename` methods from tier implementations
+
+*Contributed by [@schotime](https://github.com/schotime)*
+
+### S3 Subprocess Configuration Issues (Issue #131)
+
+Fixed compaction failures on S3-compatible storage services (Hetzner, MinIO, etc.) due to missing subprocess configuration.
+
+**Issue 1: Credentials not forwarded**
+- Subprocess fell back to AWS EC2 IMDS for credentials, timing out on non-AWS environments
+- Error: `"operation error S3: GetObject, exceeded maximum number of attempts"`
+- Fix: S3 credentials now passed via `AWS_ACCESS_KEY_ID` and `AWS_SECRET_ACCESS_KEY` environment variables
+
+**Issue 2: SSL config not forwarded**
+- Subprocess defaulted to HTTP when main process used HTTPS (port mismatch)
+- Fix: `use_ssl` config now included in subprocess configuration
+
+### Arrow Writer Panic During High-Concurrency Writes (Issue #130)
+
+Fixed a panic that occurred during high-concurrency writes when batches had different column sets (schema evolution).
+
+**Symptoms:**
+- Panic: `runtime error: index out of range [N] with length M`
+- Occurred in `sliceColumnsByIndices()` during flush operations
+- More likely with high write concurrency (many workers, large batches)
+
+**Root cause:** When batches with different schemas were merged (e.g., some records had `cpu` field, others had `temperature`), the `mergeBatches()` function created columns of different lengths. When `groupByHour()` generated indices from the `time` column, accessing those indices on shorter columns caused an out-of-bounds panic.
+
+**Fix:**
+- `mergeBatches()` now normalizes all columns to the same length, using zero values for sparse positions
+- `sliceColumnsByIndices()` now includes defensive bounds checking as an additional safety layer
+
 ### Empty Directories Not Cleaned Up After Daily Compaction
 
 Fixed an issue where empty hour-level partition directories were left behind after daily compaction consolidated files into day-level partitions.

--- a/internal/ingest/arrow_writer.go
+++ b/internal/ingest/arrow_writer.go
@@ -1888,6 +1888,7 @@ func (b *ArrowBuffer) flushBufferLockedDataTime(ctx context.Context, bufferKey, 
 
 // mergeBatches merges multiple column batches into a single columnar structure
 // OPTIMIZATION: Pre-allocate merged arrays to avoid O(n²) append reallocations
+// Handles sparse columns (schema evolution) by ensuring all columns have the same length
 func (b *ArrowBuffer) mergeBatches(batches []interface{}) (map[string]interface{}, error) {
 	if len(batches) == 0 {
 		return nil, fmt.Errorf("no batches to merge")
@@ -1901,84 +1902,90 @@ func (b *ArrowBuffer) mergeBatches(batches []interface{}) (map[string]interface{
 		return nil, fmt.Errorf("invalid batch type: %T", batches[0])
 	}
 
-	// PHASE 1: Calculate total sizes per column and detect types
+	// PHASE 1: Calculate total rows from time column and collect column type info
 	type colInfo struct {
-		totalSize int
-		colType   string // "int64", "float64", "string", "bool"
+		colType string // "int64", "float64", "string", "bool"
 	}
-	sizes := make(map[string]*colInfo)
+	colTypes := make(map[string]*colInfo)
+	totalRows := 0
 
+	// First pass: count total rows using time column
 	for _, batch := range batches {
 		cols, ok := batch.(map[string]interface{})
 		if !ok {
 			return nil, fmt.Errorf("invalid batch type: %T", batch)
 		}
 
-		for name, col := range cols {
-			info := sizes[name]
-			if info == nil {
-				info = &colInfo{}
-				sizes[name] = info
-			}
+		// Count rows from time column (always present)
+		if timeCol, ok := cols["time"].([]int64); ok {
+			totalRows += len(timeCol)
+		}
 
-			switch v := col.(type) {
-			case []int64:
-				info.totalSize += len(v)
-				info.colType = "int64"
-			case []float64:
-				info.totalSize += len(v)
-				info.colType = "float64"
-			case []string:
-				info.totalSize += len(v)
-				info.colType = "string"
-			case []bool:
-				info.totalSize += len(v)
-				info.colType = "bool"
-			default:
-				return nil, fmt.Errorf("unsupported column type: %T", v)
+		// Collect column types
+		for name, col := range cols {
+			if colTypes[name] == nil {
+				info := &colInfo{}
+				switch col.(type) {
+				case []int64:
+					info.colType = "int64"
+				case []float64:
+					info.colType = "float64"
+				case []string:
+					info.colType = "string"
+				case []bool:
+					info.colType = "bool"
+				default:
+					return nil, fmt.Errorf("unsupported column type: %T", col)
+				}
+				colTypes[name] = info
 			}
 		}
 	}
 
-	// PHASE 2: Pre-allocate merged arrays with exact capacity
-	merged := make(map[string]interface{}, len(sizes))
-	offsets := make(map[string]int, len(sizes)) // Track copy position per column
+	// PHASE 2: Pre-allocate ALL columns to totalRows (handles sparse columns)
+	merged := make(map[string]interface{}, len(colTypes))
 
-	for name, info := range sizes {
+	for name, info := range colTypes {
 		switch info.colType {
 		case "int64":
-			merged[name] = make([]int64, info.totalSize)
+			merged[name] = make([]int64, totalRows)
 		case "float64":
-			merged[name] = make([]float64, info.totalSize)
+			merged[name] = make([]float64, totalRows)
 		case "string":
-			merged[name] = make([]string, info.totalSize)
+			merged[name] = make([]string, totalRows)
 		case "bool":
-			merged[name] = make([]bool, info.totalSize)
+			merged[name] = make([]bool, totalRows)
 		}
-		offsets[name] = 0
 	}
 
-	// PHASE 3: Copy data without reallocation
+	// PHASE 3: Copy data at correct row offsets (not per-column offsets)
+	rowOffset := 0
 	for _, batch := range batches {
 		cols := batch.(map[string]interface{}) // Already validated above
 
+		// Determine batch size from time column
+		batchRows := 0
+		if timeCol, ok := cols["time"].([]int64); ok {
+			batchRows = len(timeCol)
+		}
+
+		// Copy each column's data at the current row offset
 		for name, col := range cols {
-			offset := offsets[name]
 			switch v := col.(type) {
 			case []int64:
-				copy(merged[name].([]int64)[offset:], v)
-				offsets[name] = offset + len(v)
+				copy(merged[name].([]int64)[rowOffset:], v)
 			case []float64:
-				copy(merged[name].([]float64)[offset:], v)
-				offsets[name] = offset + len(v)
+				copy(merged[name].([]float64)[rowOffset:], v)
 			case []string:
-				copy(merged[name].([]string)[offset:], v)
-				offsets[name] = offset + len(v)
+				copy(merged[name].([]string)[rowOffset:], v)
 			case []bool:
-				copy(merged[name].([]bool)[offset:], v)
-				offsets[name] = offset + len(v)
+				copy(merged[name].([]bool)[rowOffset:], v)
 			}
 		}
+		// Sparse columns that don't exist in this batch will have zero values
+		// at positions [rowOffset : rowOffset+batchRows]
+
+		rowOffset += batchRows
 	}
 
 	return merged, nil
@@ -2258,6 +2265,7 @@ func groupByHour(times []int64) (map[int64]*hourBucket, int64, int64, error) {
 
 // sliceColumnsByIndices extracts rows from all columns based on a list of indices
 // Returns a new column map with only the selected rows
+// Handles sparse columns (columns shorter than indices) by using zero values for out-of-bounds access
 func sliceColumnsByIndices(columns map[string]interface{}, indices []int) map[string]interface{} {
 	result := make(map[string]interface{}, len(columns))
 
@@ -2265,29 +2273,45 @@ func sliceColumnsByIndices(columns map[string]interface{}, indices []int) map[st
 		switch col := colData.(type) {
 		case []int64:
 			newCol := make([]int64, len(indices))
+			colLen := len(col)
 			for i, idx := range indices {
-				newCol[i] = col[idx]
+				if idx < colLen {
+					newCol[i] = col[idx]
+				}
+				// else: leave as zero value (sparse column handling)
 			}
 			result[colName] = newCol
 
 		case []float64:
 			newCol := make([]float64, len(indices))
+			colLen := len(col)
 			for i, idx := range indices {
-				newCol[i] = col[idx]
+				if idx < colLen {
+					newCol[i] = col[idx]
+				}
+				// else: leave as zero value (sparse column handling)
 			}
 			result[colName] = newCol
 
 		case []string:
 			newCol := make([]string, len(indices))
+			colLen := len(col)
 			for i, idx := range indices {
-				newCol[i] = col[idx]
+				if idx < colLen {
+					newCol[i] = col[idx]
+				}
+				// else: leave as empty string (sparse column handling)
 			}
 			result[colName] = newCol
 
 		case []bool:
 			newCol := make([]bool, len(indices))
+			colLen := len(col)
 			for i, idx := range indices {
-				newCol[i] = col[idx]
+				if idx < colLen {
+					newCol[i] = col[idx]
+				}
+				// else: leave as false (sparse column handling)
 			}
 			result[colName] = newCol
 

--- a/internal/ingest/arrow_writer_test.go
+++ b/internal/ingest/arrow_writer_test.go
@@ -491,3 +491,139 @@ func BenchmarkRowsToColumnar_SchemaVariation(b *testing.B) {
 		buffer.rowsToColumnar("cpu", rows)
 	}
 }
+
+// TestMergeBatches_SparseColumns tests that mergeBatches handles schema evolution correctly
+// This reproduces Issue #130: panic during high-concurrency writes with different column sets
+func TestMergeBatches_SparseColumns(t *testing.T) {
+	buffer := createTestArrowBuffer(t)
+
+	// Create batches with different column sets (simulating schema evolution)
+	batch1 := map[string]interface{}{
+		"time": []int64{1, 2, 3},
+		"cpu":  []float64{10.0, 20.0, 30.0},
+	}
+	batch2 := map[string]interface{}{
+		"time":        []int64{4, 5},
+		"temperature": []float64{70.0, 80.0}, // New column not in batch1
+	}
+	batch3 := map[string]interface{}{
+		"time": []int64{6, 7, 8},
+		"cpu":  []float64{40.0, 50.0, 60.0}, // cpu column returns
+	}
+
+	batches := []interface{}{batch1, batch2, batch3}
+
+	merged, err := buffer.mergeBatches(batches)
+	if err != nil {
+		t.Fatalf("mergeBatches failed: %v", err)
+	}
+
+	// Verify all columns have the same length (8 total rows)
+	expectedLen := 8
+	for colName, colData := range merged {
+		var actualLen int
+		switch col := colData.(type) {
+		case []int64:
+			actualLen = len(col)
+		case []float64:
+			actualLen = len(col)
+		case []string:
+			actualLen = len(col)
+		case []bool:
+			actualLen = len(col)
+		}
+		if actualLen != expectedLen {
+			t.Errorf("Column %q has length %d, expected %d", colName, actualLen, expectedLen)
+		}
+	}
+
+	// Verify time column has all values
+	timeCol := merged["time"].([]int64)
+	expectedTime := []int64{1, 2, 3, 4, 5, 6, 7, 8}
+	for i, expected := range expectedTime {
+		if timeCol[i] != expected {
+			t.Errorf("time[%d] = %d, expected %d", i, timeCol[i], expected)
+		}
+	}
+
+	// Verify cpu column: should have values at positions 0,1,2 and 5,6,7, zeros at 3,4
+	cpuCol := merged["cpu"].([]float64)
+	expectedCpu := []float64{10.0, 20.0, 30.0, 0.0, 0.0, 40.0, 50.0, 60.0}
+	for i, expected := range expectedCpu {
+		if cpuCol[i] != expected {
+			t.Errorf("cpu[%d] = %f, expected %f", i, cpuCol[i], expected)
+		}
+	}
+
+	// Verify temperature column: should have values at positions 3,4, zeros elsewhere
+	tempCol := merged["temperature"].([]float64)
+	expectedTemp := []float64{0.0, 0.0, 0.0, 70.0, 80.0, 0.0, 0.0, 0.0}
+	for i, expected := range expectedTemp {
+		if tempCol[i] != expected {
+			t.Errorf("temperature[%d] = %f, expected %f", i, tempCol[i], expected)
+		}
+	}
+}
+
+// TestSliceColumnsByIndices_BoundsCheck tests that sliceColumnsByIndices handles sparse columns
+func TestSliceColumnsByIndices_BoundsCheck(t *testing.T) {
+	// Create columns with different lengths (sparse scenario)
+	columns := map[string]interface{}{
+		"time":        []int64{1, 2, 3, 4, 5, 6, 7, 8},
+		"cpu":         []float64{10.0, 20.0, 30.0, 40.0, 50.0, 60.0}, // Only 6 elements
+		"temperature": []float64{70.0, 80.0},                         // Only 2 elements
+		"host":        []string{"a", "b", "c", "d"},                  // Only 4 elements
+		"active":      []bool{true, false, true},                     // Only 3 elements
+	}
+
+	// Try to slice with indices that exceed shorter columns' lengths
+	indices := []int{0, 2, 4, 6, 7} // Indices 6,7 exceed temperature, host, active columns
+
+	// This should NOT panic (previously would panic with index out of range)
+	result := sliceColumnsByIndices(columns, indices)
+
+	// Verify time column (all indices valid)
+	timeCol := result["time"].([]int64)
+	expectedTime := []int64{1, 3, 5, 7, 8}
+	for i, expected := range expectedTime {
+		if timeCol[i] != expected {
+			t.Errorf("time[%d] = %d, expected %d", i, timeCol[i], expected)
+		}
+	}
+
+	// Verify cpu column (indices 0,2,4 valid; 6 invalid -> 0.0)
+	cpuCol := result["cpu"].([]float64)
+	expectedCpu := []float64{10.0, 30.0, 50.0, 0.0, 0.0} // Last two are zero (out of bounds)
+	for i, expected := range expectedCpu {
+		if cpuCol[i] != expected {
+			t.Errorf("cpu[%d] = %f, expected %f", i, cpuCol[i], expected)
+		}
+	}
+
+	// Verify temperature column (only indices 0 valid)
+	tempCol := result["temperature"].([]float64)
+	expectedTemp := []float64{70.0, 0.0, 0.0, 0.0, 0.0} // Only first is valid
+	for i, expected := range expectedTemp {
+		if tempCol[i] != expected {
+			t.Errorf("temperature[%d] = %f, expected %f", i, tempCol[i], expected)
+		}
+	}
+
+	// Verify host column (indices 0,2 valid)
+	hostCol := result["host"].([]string)
+	expectedHost := []string{"a", "c", "", "", ""} // Indices 4,6,7 out of bounds
+	for i, expected := range expectedHost {
+		if hostCol[i] != expected {
+			t.Errorf("host[%d] = %q, expected %q", i, hostCol[i], expected)
+		}
+	}
+
+	// Verify active column (indices 0,2 valid)
+	activeCol := result["active"].([]bool)
+	expectedActive := []bool{true, true, false, false, false} // Indices 4,6,7 out of bounds
+	for i, expected := range expectedActive {
+		if activeCol[i] != expected {
+			t.Errorf("active[%d] = %v, expected %v", i, activeCol[i], expected)
+		}
+	}
+}

